### ISSUE TITLE
Review of plugin

### ISSUE THREAD

**Issue by pamtbaau** [X-Ryl669/grav-plugin-viewer3-d#1]
@X-Ryl669, I've taken the liberty to test-run your plugin. In short: I cannot get it to work...

I proceeded as follows:
- `$ cd user/plugins`
- `$ mkdir viewer3-d`
- `$ cd viewer3-d`
- `$ git clone https://github.com/X-Ryl669/grav-plugin-viewer3-d .`
- Created page `user/pages/03.viewer/default.md`
- Downloaded `https://blog.cyril.by/en/documentation/3d%20models/filament-detector/SupportFilDet.stl` into folder
- Added the following inside the markdown of the page:
  <pre>
  ```3dv
  SupportFilDet.stl
  ```
  </pre>
- Pointed browser to `localhost/grav/viewer`
  - No 3D is being shown
  - A `<div class="online_3d_viewer"...>` is being created, containing two other divs but all are empty.

Other remarks:
- I would suggest using a Shortcode instead of Markdown extension:
The plugin is extending the block code fences (triple back-ticks) using <code>```3dv</code> to generate the code to view the 3D model. IMHO adding a 3D viewer to the page is not really what block code fences are meant to be used for.
- I think a name like "viewer-3d" would be a more suitable name. 
- In its current state, I don't think the plugin should be released to the community. My suggestion would be to close this issue and open a new one once the plugin is ready for production.

**Comment by X-Ryl669** [comment]
So, do you have a javascript error in your browser's console ?

As I said in the other issue, the node `<div class="online..."` is written with 2 children node at once. From there, you should also have 3 javascript files added to your page (`three.min.js`, `o3dv.min.js` and `3dviewer.js`). If they aren't loaded or if they fails, it should appear in the Javascript console or the network tab.

**Comment by X-Ryl669** [comment]
(Sorry for misreading your description)

**Comment by X-Ryl669** [comment]
Do you have jquery on your website ? The plugin requires jquery, but since it's loaded by default in my Grav installation, I haven't included it in the required resources. I suspect it would fail to load if you don't have it in your default installation.

Said differently, does it work when you go [here](https://blog.cyril.by/en/documentation/3d%20models/filament-detector) ?

**Comment by pamtbaau** [comment]
The viewer isn't created because assets are never being loaded: `onTwigSiteVariables` is fired before `onMarkdownInitialized`. See https://learn.getgrav.org/17/plugins/grav-lifecycle

**Comment by X-Ryl669** [comment]
That's not what written [here](https://learn.getgrav.org/17/plugins/event-hooks) and it works on my site.
In the referred page, `onMarkdownInitialized` is 19.2th and `onTwigSiteVariables` is 25th. 
In your referred page, `onTwigSiteVariables` (do not confuse with `onTwigInitialized`) seems to happen *after* running the page processor. The page processor must render markdown (althrough there is no description as to when page.content() is called), I suppose it's done *after* `onTwigInitialized` but before `onTwigSiteVariables`

**Comment by X-Ryl669** [comment]
You can put a `var_dump("markdown")` in `onMarkdownInitialized` and `var_dump("twig")` in `onTwigSiteVariables` to check in which order they are called. On my website, I'm getting:
```
string(8) "markdown"
string(16) "twigsitevariable"
```
So it's clearly behaving like expected for me.

**Comment by pamtbaau** [comment]
Yes, I believe it is working in your env... 

Would you mind taking a step back and pretend to be a new user:
- Install fresh download of Grav 1.7.14
- Download zip version of repo
- Unzip into `user/plugins`
- Rename folder to `viewer3-d`
- Create page `user/pages/03.viewer/default.md`
- Download `https://blog.cyril.by/en/documentation/3d%20models/filament-detector/SupportFilDet.stl` into folder
- Add the following inside the markdown of the page:
  <pre>
  ```3dv
  SupportFilDet.stl
  ```
  </pre>
- Point browser to `localhost/grav/viewer`

If I perform above steps:
- no 3d model is being shown, although `<div class="online_3d_viewer"...>` is being created.
- when page is NOT yet cached, according the debugger and var_dump: `onTwigSiteVariables` is run before `onMarkdownInitialized`
- when page has been cached, according the debugger and var_dump: only `onTwigSiteVariables` is run.

**Comment by X-Ryl669** [comment]
That's strange. I've started a website like you said and it behaves differently as you said. I don't understand why, but clearly the documentation is wrong here. In the documentation page [here](https://learn.getgrav.org/17/plugins/event-hooks) it says that `onMarkdownInitialized` happens in `onPageInitialized` (unconditionally off the cached state of the page) before `onTwigSiteVariables`. This is also what's stated in your linked [documentation page](https://learn.getgrav.org/17/plugins/grav-lifecycle): `onPageInitialized` happens in the page processor (position 10.3), while `onTwigSiteVariables` *should* happen later in position 12.3.1.

I think it's a bug with grav or its documentation. 
Else, no one can load additional assets required by markdown extensions **conditionally** (short code or markdown, I don't think it would make a change here). 

Obviously I could force loading the assets anywhere (replace the line that read `if (isset($this->used) && $this->used)` by `if (true)` and it'll work, but that's poor solution since it forces loading threejs library that's huge for all the website's pages.

So, I guess I'll have to fill a bug in Grav's repository here or maybe you know why it's not following the documented lifecycle ?

**Comment by X-Ryl669** [comment]
BTW, it's also requiring to add a media type for STL files in your site, typically like this (in Configuration / Media tab):
![image](https://user-images.githubusercontent.com/3277165/116912322-58a89c80-ac48-11eb-9dff-263f8816e08f.png)

**Comment by pamtbaau** [comment]
Glad you've done the test... I wanted you see for your self which 3 issues (js assets, images, media) were blocking the display of the viewer.

Yes, page [Event hooks](https://learn.getgrav.org/17/plugins/event-hooks) is misleading and I'm currently in the process of rewriting it.
[Grav lifecycle]( https://learn.getgrav.org/17/plugins/grav-lifecycle) is correct and more clear. 

You may be misconceiving the aspect of caching used by Grav. Read the bottom of [Grav lifecycle]( https://learn.getgrav.org/17/plugins/grav-lifecycle) 

- If content is NOT cached:
   - Fire onPageContentRaw event
   - Process the page according to Markdown and Twig settings. Fire onMarkdownInitialized event
   - Fire onPageContentProcessed event
- Fire onPageContent event

Meaning:
- `onMarkdownInitialized`, is only called once, when the page has not yet been cached, or the cache has been invalidated. That means that after the first run, your extension won't run again and `$this->used` will not be set again.
- A 'trick' to circumvent that is by setting `cache_enable: false` in the page on which the viewer needs to be shown.
Because your extension will now be fired every time the page with the viewer is being requested, you can use any event after `onMarkdownInitialized`, but before `onOutputGenerated`.
Of course, this defies Grav's strength of caching. See below on Shortcodes.

Besides the loading of javascript, two other roadblocks were preventing the viewer to be shown:
- The location of the images, which you have fixed.
-  The loading of the the model file for which two solutions are possible:
   - Changing the `media.yaml` file
   - Or changing `model='" .$slug . '/' .$Block['src']` to the path of the file e.g. `/user/pages/03.viewer/SupportFilDet.stl`. In this case, the file is loaded without intervention of Grav and no `media.yaml` change is needed.

Final suggestion: Have a look at the [Shortcode Core](https://github.com/getgrav/grav-plugin-shortcode-core) plugin. That is the Grav way of doing it. When using Shortcodes, the `cache_enable` trick is no longer needed and the page's content can be efficiently cached. Also, assets are being cached. Quite more efficient...

**Comment by X-Ryl669** [comment]
Thanks for your hints, it's much appreciated! 

I think changing the `media.yaml` is the only way that would work, since the plugin allows to download the model and as such, the correct mime-type must be set for downloading it. So if I link to the file directly, I think (not tested), that the http server will not send the correct mime-type by itself (I don't think nginx or apache have the mime-type for all supported 3d model file).

Can I change it automatically in my plugin (instead of asking the user to do it manually) ?

How can I specify in my plugin that it depends on the Shortcode Core plugin ?

**Comment by pamtbaau** [comment]
> How can I specify in my plugin that it depends on the Shortcode Core plugin ?

Inside `blueprint.yaml`, add something like the following to the `dependencies` section:
```
- { name: shortcode-core, version: '>=5.0.0' }
```

> So if I link to the file directly, I think (**not tested**)

Test it and you'll know it. "_The proof is in the pudding..._" (not the mind)

**Comment by X-Ryl669** [comment]
> Test it and you'll know it

Tested and it does not work. Got `Content-Type: application/vnd.ms-pki.stl` and not `model/stl` as expected. I don't know where the former is coming from (I don't manage the http server on my test server), but it's clearly incorrect.

**Comment by X-Ryl669** [comment]
I'm closing this now, since I'll remove this repository in favor of [shortcode's version here](https://github.com/X-Ryl669/grav-plugin-viewer3d)
